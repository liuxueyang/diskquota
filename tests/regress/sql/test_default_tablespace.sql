-- test role_tablespace_quota works with tables/databases in default tablespace
-- test role_tablespace_quota works with tables/databases in non-default tablespace with hard limits on

-- start_ignore
\! mkdir -p /tmp/custom_tablespace
-- end_ignore

DROP ROLE if EXISTS role1;
CREATE ROLE role1 SUPERUSER;
SET ROLE role1;

DROP TABLE if EXISTS t;
CREATE TABLE t (i int);

-- with hard limits off
SELECT diskquota.disable_hardlimit();
SELECT diskquota.set_role_tablespace_quota('role1', 'pg_default', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to success
INSERT INTO t SELECT generate_series(1, 100);
INSERT INTO t SELECT generate_series(1, 1000000);
-- expect insert to fail
INSERT INTO t SELECT generate_series(1, 1000000);

DROP TABLE IF EXISTS t;
CREATE TABLE t (i int);

-- with hard limits on
SELECT diskquota.enable_hardlimit();
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to fail
INSERT INTO t SELECT generate_series(1, 1000000);
DROP TABLE IF EXISTS t;

-- database in customized tablespace
CREATE TABLESPACE custom_tablespace LOCATION '/tmp/custom_tablespace';
CREATE DATABASE db_with_tablespace TABLESPACE custom_tablespace;
\c db_with_tablespace;
SET ROLE role1;
CREATE EXTENSION diskquota;

-- with hard limits off
SELECT diskquota.disable_hardlimit();
SELECT diskquota.set_role_tablespace_quota('role1', 'custom_tablespace', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert to success
CREATE TABLE t_in_custom_tablespace AS SELECT generate_series(1, 100);
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);
-- start_ignore
SELECT * from diskquota.blackmap;
-- end_ignore
-- expect insert to fail
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);

-- with hard limits on
SELECT diskquota.enable_hardlimit();
SELECT diskquota.wait_for_worker_new_epoch();

DROP TABLE IF EXISTS t_in_custom_tablespace;
-- expect create table to fail
CREATE TABLE t_in_custom_tablespace AS SELECT generate_series(1, 1000000);
-- start_ignore
SELECT * from diskquota.blackmap;
-- end_ignore

-- clean up
DROP TABLE IF EXISTS t_in_custom_tablespace;
SELECT diskquota.disable_hardlimit();
DROP EXTENSION IF EXISTS diskquota;
\c contrib_regression;
DROP DATABASE IF EXISTS db_with_tablespace;
DROP TABLESPACE IF EXISTS custom_tablespace;

RESET ROLE;
DROP ROLE IF EXISTS role1;

SELECT diskquota.disable_hardlimit();
