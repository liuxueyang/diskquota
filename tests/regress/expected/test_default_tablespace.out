-- test role_tablespace_quota works with tables/databases in default tablespace
-- test role_tablespace_quota works with tables/databases in non-default tablespace with hard limits on
-- start_ignore
\! mkdir -p /tmp/custom_tablespace
-- end_ignore
DROP ROLE if EXISTS role1;
NOTICE:  role "role1" does not exist, skipping
CREATE ROLE role1 SUPERUSER;
SET ROLE role1;
DROP TABLE if EXISTS t;
NOTICE:  table "t" does not exist, skipping
CREATE TABLE t (i int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
-- with hard limits off
SELECT diskquota.disable_hardlimit();
 disable_hardlimit 
-------------------
 
(1 row)

SELECT diskquota.set_role_tablespace_quota('role1', 'pg_default', '1 MB');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to success
INSERT INTO t SELECT generate_series(1, 100);
INSERT INTO t SELECT generate_series(1, 1000000);
-- expect insert to fail
INSERT INTO t SELECT generate_series(1, 1000000);
ERROR:  tablespace:pg_default role:role1 diskquota exceeded
DROP TABLE IF EXISTS t;
CREATE TABLE t (i int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
-- with hard limits on
SELECT diskquota.enable_hardlimit();
 enable_hardlimit 
------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to fail
INSERT INTO t SELECT generate_series(1, 1000000);
ERROR:  tablespace:1663 role:3583344 diskquota exceeded  (seg0 127.0.0.1:6002 pid=26100)
DROP TABLE IF EXISTS t;
-- database in customized tablespace
CREATE TABLESPACE custom_tablespace LOCATION '/tmp/custom_tablespace';
CREATE DATABASE db_with_tablespace TABLESPACE custom_tablespace;
\c db_with_tablespace;
SET ROLE role1;
CREATE EXTENSION diskquota;
-- with hard limits off
SELECT diskquota.disable_hardlimit();
 disable_hardlimit 
-------------------
 
(1 row)

SELECT diskquota.set_role_tablespace_quota('role1', 'custom_tablespace', '1 MB');
 set_role_tablespace_quota 
---------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert to success
CREATE TABLE t_in_custom_tablespace AS SELECT generate_series(1, 100);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column(s) named 'generate_series' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);
-- start_ignore
SELECT * from diskquota.blackmap;
      target_type      | target_oid | database_oid | tablespace_oid | seg_exceeded | dbnode | spcnode | relnode | segid 
-----------------------+------------+--------------+----------------+--------------+--------+---------+---------+-------
 ROLE_TABLESPACE_QUOTA |    3583344 |      3583352 |        3583351 | f            |      0 |       0 |       0 |    -1
(1 row)

-- end_ignore
-- expect insert to fail
INSERT INTO t_in_custom_tablespace SELECT generate_series(1, 1000000);
ERROR:  tablespace:custom_tablespace role:role1 diskquota exceeded
-- with hard limits on
SELECT diskquota.enable_hardlimit();
 enable_hardlimit 
------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

DROP TABLE IF EXISTS t_in_custom_tablespace;
-- expect create table to fail
CREATE TABLE t_in_custom_tablespace AS SELECT generate_series(1, 1000000);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column(s) named 'generate_series' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
ERROR:  tablespace:custom_tablespace role:role1 diskquota exceeded
-- start_ignore
SELECT * from diskquota.blackmap;
      target_type      | target_oid | database_oid | tablespace_oid | seg_exceeded | dbnode | spcnode | relnode | segid 
-----------------------+------------+--------------+----------------+--------------+--------+---------+---------+-------
 ROLE_TABLESPACE_QUOTA |    3583344 |      3583352 |        3583351 | f            |      0 |       0 |       0 |    -1
(1 row)

-- end_ignore
-- clean up
DROP TABLE IF EXISTS t_in_custom_tablespace;
NOTICE:  table "t_in_custom_tablespace" does not exist, skipping
SELECT diskquota.disable_hardlimit();
 disable_hardlimit 
-------------------
 
(1 row)

DROP EXTENSION IF EXISTS diskquota;
\c contrib_regression;
DROP DATABASE IF EXISTS db_with_tablespace;
DROP TABLESPACE IF EXISTS custom_tablespace;
RESET ROLE;
DROP ROLE IF EXISTS role1;
SELECT diskquota.disable_hardlimit();
 disable_hardlimit 
-------------------
 
(1 row)

